### PR TITLE
More race-condition safe ticket/datetime reserved/sold updates

### PR DESCRIPTION
**Pull request by mnelson4**
## Problem this Pull Request solves

Added more race-condition safe ticket/datetime reserved/sold updates to avoid overselling.
Addresses https://github.com/eventespresso/event-espresso-core/issues/306.
Instead of reading the ticket or datetime's reserved or sold counts, updating them on the model object, then saving the entire model object to the DB, we now issue DB queries that only increment or decrement the specific column's value directly to the DB, and afterwards update the model object. This will reduce the likelihood of two simultaneous requests overwriting each other's modifications to the ticket and datetimes' reserved and sold counts.

## How has this been tested

* [x] Submit the ticket selector form, but don't fill out any attendee info. Verify the ticket's reserved count goes up appropriately.
* [x] Press back in your browser and re-submit the ticket selector with a different amount. Master's behaviour should be preserved: your old tickets were released and your new ticket selections have been reserved.
* [x] Now finish the registration. Again the behaviour should be identical to master: your tickets are no longer reserved, but sold. Verify the ticket's sold and reserved counts are accurate.
* [x] Now add this code snippet: 
```
add_action(
    'AHEE__EE_Ticket__increase_reserved__begin',
    function(
        EE_Ticket $ticket,
        $quantity,
        $source
    ) {
        global $wpdb;
        // simulate someone swooping in and reserving a ticket before us
        $success = EEM_Ticket::instance()->update(
            [
                'TKT_reserved' => $ticket->reserved() + 3
            ],
            [
                [
                    'TKT_ID' => $ticket->ID()
                ]
            ],
            false
        );
    },
    10,
    3
);
```
to your functions.php. It simulates a simultaneous request coming in which reserves 3 tickets. Create a new ticket. Select 1 ticket on the ticket selector. The ticket's reserved count should now be 4. (Your 1, plus 3 which the code snippet simulated). Compare this to the behaviour on `FET/conditional-update-queries-only-add-action`, which is a branch of master which just adds a single action to facilitate simulating a simultaneous request. On that branch, the simulated ticket purchase gets overwritten, so the ticket's reserved count would have been 1 (which is wrong given there was a simulated simultaneous request that bumped it up to 3, so it should have been 4 like on this PR's branch.) Once you're done this testing step, remove that code snippet.

 ### testing release of reserved tickets for abandoned carts

- start by adding the following filter to an appropriate location in your testing site (functions.php, etc)

```php
// reduce 60 minute session lifespan to 5 minutes
add_filter(
    'FHEE__EE_Session__construct___lifespan',
    function($lifespan) {
        return $lifespan / 12;
    }
);
```
- the above will set the session lifespan to 5 minutes, but feel free to shorten the session lifespan even more (dividing by 15 will give you a 4 minute session, dividing by 20 will give you a 3 minute session, dividing by 30 will give you a 2 minute session, etc) 
- pick an active site with available tickets and visit its event editor in the admin
- take note of the sold and reserved counts for the ticket and datetime(s) you plan to select
- in another browser tab or window, visit that event on the frontend and select one or more of your chosen ticket
- proceed to registration but stop at the first step
- switch to the event editor in your other browser tab and refresh the page
- [x] confirm that the reserved counts for the ticket and datetime(s) you selected have been incremented by the correct amounts
- now close the other browser tab/window with the registration in progress and wait 5 minutes (or whatever your session lifespan is)
- refresh the event editor page again
- [x] confirm that the reserved counts for the ticket and datetime(s) you had selected have returned to their original values
- repeat the above process but abandon the registration process at different stages including doing so from an offsite payment gateway like PayPal standard.
- [x] confirm that the reserved counts for selected tickets and datetimes always increase correctly then return to their original values after your session has expired
- experiment with sessions and reserved counts in this branch and attempt to find any potential issues
- plz compare any potential issues against master branch to determine if they are caused by the work in this branch

 ### testing effects of ticket and datetime relations in the event editor

- log into the admin and go to an event that already has some tickets and datetimes with existing sales (non-zero sold counts)
- experiment with duplicating tickets and datetimes
- [x] confirm that sold and reserved counts for duplicated tickets and datetimes are set to zero and existing values are not affected
- select an original ticket that has a non-zero sold count, click the gear icon to view its advanced settings, and assign it to one of the duplicated datetimes with a zero sold count
- save the event
- [x] confirm that the sold count for the duplicated datetime has increased by the sold count of the added ticket
- now remove that same ticket from one of the datetimes it was originally assigned to and resave the event
- [x] confirm that the sold count for the original datetime has decreased by the sold count of the removed ticket
- experiment with sold and reserved counts in the editor in this branch and attempt to find any potential issues
- plz compare any potential issues against master branch to determine if they are caused by the work in this branch

## Checklist

* [x] I have read the documentation relating to systems affected by this pull request, see https://github.com/eventespresso/event-espresso-core/tree/master/docs
* [x] User input is adequately validated and sanitized
* [x] all publicly displayed strings are internationalized (usually using `esc_html__()`, see https://codex.wordpress.org/I18n_for_WordPress_Developers)
* [x] My code is tested.
* [x] My code follows the Event Espresso code style.
* [x] My code has proper inline documentation.
* [x] My code accounts for when the site is in Maintenance Mode (MM2 especially disallows usage of models)

### CODE REVIEW

**Comment by tn3rb** [comment]
> I think we forgot about that in a few places where we intended to only update the sold counts. Specifically:
> 
> - `espresso_events_Pricing_Hooks::_update_ticket_datetimes()`
> - `EE_Ticket::update_tickets_sold()`
> - `EE_Datetime::update_sold()`

hmmmm... hard to say what _should_ happen with those last two since they were added as an attempt to make the sold counts match the actual registration counts. ie: if the sold count says 7 but there are actually 6 registrations (or 8 registrations)... what should the reserved count be?

**HOWEVER**...

`espresso_events_Pricing_Hooks::_update_ticket_datetimes()` **IS** definitely a bug and could be one of the ways that events are ending up with inaccurate sold counts. AFAIK it has never been reported that the inaccuracies only occur after duplicating tickets or datetimes, but that may not get noticed until an event sells out. So event admins not making that connection is easily explainable.

That said, this would only be capable of borking the counts if an event admin was duplicating tickets and/or datetimes within an hour of the last tickets being sold (or whatever amount of time the session expiration is set to), because otherwise the inaccurately low reserved count would just wind down to zero as sessions expired (or unless there were continuous registrations right up until the sell out occurs so that the reserved count never reaches zero). So even though this is a potential cause of inaccuracies, it would be unlikely to occur and rare at best imho. A busy site with our existing queries (prior to this branch) is far more likely to cause problems and better fits the reported data.

anyways... good find.

**Comment by mnelson4** [comment]
> mmmm... hard to say what should happen with those last two since they were added as an attempt to make the sold counts match the actual registration counts.

Upon further reflection, I think this is a definitely possible source of overselling. Consider this:
- a ticket has a limit of 10 tickets
- someone begins registering for 9, but doesn't complete
- the admin resets the ticket sold counts (brings reserved back down to 0)
- in another tab/browser, someone registers for 2 or more tickets and completes (sold is now at 2)
- the original registration for 9 completes (sold is now at 2 + 9 = 11)

I doubt that's usually the problem. But it's a possible source of problems...

Is there a way to calculate the reserved count (external to what's in the datetime and tickets's reserved counts)?

> espresso_events_Pricing_Hooks::_update_ticket_datetimes() IS definitely a bug and could be one of the ways that events are ending up with inaccurate sold counts. 

I played with it for a while and couldn't figure out how to oversell because of this. But ya, it's fishy so I'm inclined to just set the sold count and leave the reserved count unaffected.

**Comment by tn3rb** [comment]
> Upon further reflection, I think this is a definitely possible source of overselling...
> ...
> - the admin resets the ticket sold counts...

oh ya manually resetting the sold counts can definitely mess with the counts but I don't think we should consider that since (I'm assuming) admins are only doing that because:

- their sold counts are already completely off due to other failed mechanisms
- their event(s) are all over for the season / time being, and they want to start with a clean slate

> I doubt that's usually the problem. But it's a possible source of problems...

I totally agree 💯 

> Is there a way to calculate the reserved count (external to what's in the datetime and tickets's reserved counts)?

There is currently no mechanism in place to do so, and it wouldn't be too too hard to calculate, but I think that doing so, and then attempting to reset those values in the db, could easily be problematic. I'll explain:

- event spaces are reserved as soon as tickets have been successfully added to the cart after submitting the ticket selector form. This is represented in the database by incrementing reserved counts for each ticket and any of its associated datetimes.
- reservations are only good for as long as the session is active. This gives users enough time to complete the registration process without the event selling out from under them, including enough time to visit off-site payment gateways.
- reservations are released for the following reasons:
    - the session has expired and the user did not complete the registration process for whatever reason
    - the user completed the registration process for a non-free event but deferred payment until a later date (invoice, etc)
    - the user completed the registration process in full including 100% payment for non-free events

So calculating the reserved counts would require knowing the number of "in-progress" registrations for each ticket, and in order to get that information we would need to know the following:
    - all active sessions - not difficult now, but far easier if we implement a custom session table
    - all registrations for each active session - easy, since each registration stores the SSN ID
    - whether the transaction is in progress ie: not finalized - easy, just call `EE_Transaction::all_reg_steps_completed()`

This requires a few queries and a bit of processing and **easily** results in a race condition if you want to try and "correct" these numbers while any registrations are in progress. In other words, by the time you calculate the "correct" number of reservations for a ticket, that number could become stale due to activity on the site, so you wouldn't want to replace what's in the db with the stale value.

Now the only way you could confidently calculate and correct what's in the db is by doing so during a period of zero registration process activity. This could be achieved via the following:

- setting a cron to run during a time that is normally not active for registrations such as the early morning, ex: 3-4 am
- having a way to monitor registrations in progress so as to be 100% certain that no registrations are under way
- temporarily suspending access to the registration process / ticket sales via something like maintenance mode - maybe a new level that allows all other event related activity (like viewing events) to proceed, but does not allow access to the ticket selector or checkout
- at this point, there should be:
    - no active sessions and therefore no registrations in progress
    - no chance for any new request to initiate the registration process
- this means that **ALL** ticket and datetime reservation counts should be zero since reservations are only held during the registration process. Therefore no calculation is required and it should be safe to simply reset all ticket and datetime reservation counts in the db to zero.

<hr>

Going back to discussing methods that update sold counts and potentially throw off reserved counts, I looked into their usages and found the following:

 ### EE_Ticket::update_tickets_sold()

This is actually only called by `EEM_Ticket::update_tickets_sold()` which in turn is only called by `EE_Import::save_csv_data_array_to_db()`. I thought that there was a mechanism somewhere in the event editor save process that also called this method but that does not appear to be the case. There could be some other mechanism in the admin that is doing this via other methods however.

 ### EE_Datetime::update_sold()

This is only called by `EEM_Datetime::update_sold()` which immediately throws an `EE_Error::doing_it_wrong()` notice instructing devs to use `EE_Ticket::update_tickets_sold()` instead. No other code appears to be calling this method.

Further to the above:

 ### EE_Ticket::set_sold()

Currently only called by: 

- `EE_Ticket::increase_sold()`
- `EE_Ticket::decrease_sold()`
- `espresso_events_Pricing_Hooks::_duplicate_ticket()` which sets the value on the duplicated ticket to zero

which is great, but the same thing could be achieved by calling `EEM_Base::set()` with `TKT_sold` for the field name. This would result in sold counts being changed without a corresponding change in the reserved counts (if required). This potential discrepancy could be tightened up by overriding `EEM_Base::set()` in `EEM_Ticket`, detecting attempts to directly modify sold counts, and then redirecting those calls to `EE_Ticket::increase_sold()` or `EE_Ticket::decrease_sold()`. This same idea could be applied to uses of `EEM_Base::set()` for `TKT_reserved` as well as `DTT_sold` and `DTT_reserved` on `EEM_Datetime`.

File: core/db_classes/EE_Ticket.class.php [hunk 1]
```diff
@@ -859,62 +865,114 @@ public function set_reserved($reserved)
...
+                    } else {
+                        $limit_exceeded = true;
+                        break;
+                    }
                 }
             }
+            // If somewhere along the way we detected a datetime whose
+            // limit was exceeded, do a manual rollback.
+            if( $limit_exceeded ) {
+                $this->decreaseReservedForDatetimes($datetimes_updated, $qty);
```
Comment (tn3rb) [inline core/db_classes/EE_Ticket.class.php]: noice 👍

File: core/db_classes/EE_Ticket.class.php [hunk 2]
```diff
@@ -923,48 +981,49 @@ protected function _increase_reserved_for_datetimes($qty = 1)
...
+            [
+                'TKT_reserved' =>  absint($qty) * -1
+            ]
+        );
+        do_action(
+            'AHEE__EE_Ticket__decrease_reserved',
+            $this,
+            $qty,
+            $reserved
+        );
```
Comment (tn3rb) [inline core/db_classes/EE_Ticket.class.php]: needs to return `true` no?

File: core/db_classes/EE_Base_Class.class.php [hunk 3]
```diff
@@ -3054,6 +3054,192 @@ public function refresh_from_db()
...
+        if(empty($fields)){
+            throw new InvalidArgumentException(esc_html__('EE_Base_Class::updateFieldsInDB was passed an empty array of fields.', 'event_espresso'));
+        }
+        $first_field = reset($fields);
+        $table_alias = $first_field->get_table_alias();
+        foreach($fields as $field) {
+            if( $table_alias !== $field->get_table_alias()) {
+                throw new InvalidArgumentException(
+                    sprintf(
+                        esc_html__('EE_Base_Class::updateFieldsInDB was passed fields for different tables ("%1$s" and "%2$s"), which is not supported. Instead, please call the method multiple times.', 'event_espresso'),
```
[the remark this answers is not in the document]
Comment (mnelson4) [reply core/db_classes/EE_Base_Class.class.php]: :+1:

File: core/db_classes/EE_Base_Class.class.php [hunk 4]
```diff
@@ -3054,6 +3054,192 @@ public function refresh_from_db()
...
+     *                                   on the same table (eg don't pass in one field for the posts table and
+     *                                   another for the event meta table.)
+     * @return bool
+     * @throws EE_Error
+     * @throws InvalidArgumentException
+     * @throws InvalidDataTypeException
+     * @throws InvalidInterfaceException
+     * @throws ReflectionException
+     */
+    public function bump(array $fields_n_quantities)
```
[the remark this answers is not in the document]
Comment (mnelson4) [reply core/db_classes/EE_Base_Class.class.php]: hmmm like you point out later, it's handy being able to increase one field whilst decreasing another one. So I'd prefer to keep the functionality as-is (name is up for debate, IMO `adjust` or `adjustInDB` would also be fairly intuitive)

File: core/db_classes/EE_Registration.class.php [hunk 5]
```diff
@@ -163,7 +163,7 @@ public function set_status($new_STS_ID = null, $use_default = false, ContextInte
             // TO approved
             if ($new_STS_ID === EEM_Registration::status_id_approved) {
                 // reserve a space by incrementing ticket and datetime sold values
-                $this->_reserve_registration_space();
+                $this->_sell_registration_space();
```
[the remark this answers is not in the document]
Comment (mnelson4) [reply core/db_classes/EE_Registration.class.php]: kk

File: core/db_classes/EE_Ticket.class.php [hunk 6]
```diff
@@ -745,92 +747,96 @@ public function set_sold($sold)
...
 
 
-    /**
-     * Decreases sold on related datetimes
-     *
-     * @param int $qty
-     * @return void
-     * @throws \EE_Error
-     */
-    protected function _decrease_sold_for_datetimes($qty = 1)
```
[the remark this answers is not in the document]
Comment (mnelson4) [reply core/db_classes/EE_Ticket.class.php]: kk, I just highly doubt anyone has overrideen `EE_Ticket`. We tried to once and things blew up. But no harm decrementing the way you did.

**Comment by mnelson4** [comment]
> hmmm... I'm unable to assign this to you for review as your name does not appear in the list of potential reviewers...
I think I can't be a reviewer when I made the original PR. 


> This requires a few queries and a bit of processing and easily results in a race condition if you want to try and "correct" these numbers while any registrations are in progress.

Ugh let's just leave it as-is. If we can only recalculate the reserved count accurately while no tickets are reserved, it's not of much use anyway. Or at least that can be a separate ticket.

> overriding EEM_Base::set() in EEM_Ticket, detecting attempts to directly modify sold counts, and then redirecting those calls to EE_Ticket::increase_sold() or EE_Ticket::decrease_sold().

That would be kinda nice but I'm worried about it. Anywhere we have code calling `EE_Base_Class::set()` they'll need to beware they might be setting a ticket/datetimes reserved/sold counts, which would cause unexpected DB queries or maybe even an accidental code loop.

On that topic, I actually just made a commit to USE those methods in `EE_Ticket::update_tickets_sold()` and `EE_Datetime::update_sold()` seeing how there was already a race condition in them, and those already shouldn't be called during hot registrations.

And about naming `bump`... I'd also go for `adjust` or `adjustInDB` or something like that. But ya, it would be nice to not change it ot `increase` or `decrease` because it can increase one field while simultaneously decreasing another.

And yes, `bumpConditionally()` actually only works on positive amounts, so it could be renamed to `increaseConditionally()` or something. 

If this is looking good, could you update the testing notes? I'm not sure if you saw it already but I asked if you had tips on how to test two affected features.

**Comment by tn3rb** [comment]
I'm ok with keeping `bump()` as is as long as we rename the other method to something like `incrementConditionally()` to indicate that it can not be used to decrease a field's value, although I still think that something really explicitly descriptive like `adjustNumericFieldValue()` would be more instantly understandable than `bump()` which conveys little meaning if you don't already know what it does. BUt I'm not going to push that any further and will leave it up to you at this point.

I simply forgot about updating the testing notes after all of the other discussions, so will do that right now

**Comment by joshfeck** [comment]
I started testing this and found right away that the reserved counts **for datetimes** do not reset back to zero after the transaction is completed with payment. 
screen shot 2019-01-18 at 4 10 55 pm
screen shot 2019-01-18 at 4 11 47 pm

**Comment by mnelson4** [comment]
Dang, yep, I can confirm this... Investigating why...

**Comment by mnelson4** [comment]
A boolean was backwards. Flipped it the right way around and added some tests to prevent regression.
@tn3rb please review again.

File: core/db_classes/EE_Datetime.class.php [hunk 7]
```diff
@@ -254,16 +254,16 @@ public function set_sold($sold)
     public function increaseSold($qty = 1, $also_decrease_reserved = true)
     {
         $qty = absint($qty);
-        if( $also_decrease_reserved) {
+        if ($also_decrease_reserved) {
             $success = $this->adjustNumericFieldsInDb(
                 [
+                    'DTT_reserved' => $qty * -1,
```
Comment (tn3rb) [inline core/db_classes/EE_Datetime.class.php]: doh... ya i totally missed that too!

**Comment by mnelson4** [comment]
passing off to @joshfeck for testing, but even if it passes testing, y'all might want to hold off on merging until all my baby issues are settled (when I'll be available to respond quickly to bug reports if any) which will probably be next week

**Comment by mnelson4** [comment]
Ok baby's here and my schedule is returning to normal, so I say testing can proceed and release if it all checks out ok.

**Comment by joshfeck** [comment]
@mnelson4 can you resolve the merge conflict with master branch? It looks like there was another branch that changed a method name.

**Comment by mnelson4** [comment]
kk conflict resolved. If the unit tests pass fully, testing can proceed.

**Comment by joshfeck** [comment]
This is tested and ready to be merged.

**Pull request merged**